src/stim/simulators/frame_simulator.pybind.cc

```diff
@@ -770,4 +770,102 @@ void stim_pybind::pybind_frame_simulator_methods(
                 [stim.PauliString("+YX__")]
         )DOC")
             .data());
+
+    c.def(
+        "broadcast_pauli_errors",
+        [](FrameSimulator<MAX_BITWORD_WIDTH> &self,
+           const pybind11::object &pauli,
+           const pybind11::object &mask
+           ) {
+            uint8_t p = 255;
+            try {
+                p = pybind11::cast<uint8_t>(pauli);
+                if (p >= 4) {
+                    throw pybind11::cast_error();
+                }
+            } catch (const pybind11::cast_error &) {
+                try {
+                    std::string s = pybind11::cast<std::string>(pauli);
+                    if (s == "X") {
+                        p = 1;
+                    } else if (s == "Y") {
+                        p = 2;
+                    } else if (s == "Z") {
+                        p = 3;
+                    } else if (s == "I" || s == "_") {
+                        p = 0;
+                    } else {
+                        throw pybind11::cast_error();
+                    }
+                } catch (const pybind11::cast_error &) {
+                    throw std::invalid_argument(
+                    "broadcast_pauli_errors only accepts pauli arguments in ['I', '_', 'X', 'Y', 'Z', 0,1,2,3]");
+                }
+            }
+
+            bool flip_z_part = p & 2;
+            bool flip_x_part = (0b0110 >> p) & 1;  // parity of 2 bit number
+
+
+            if (!pybind11::isinstance<pybind11::array_t<bool>>(mask)) {
+                throw std::invalid_argument(
+                "broadcast_pauli_errors can only accept mask that is a 2D array of np.bool_");
+            }
+            const pybind11::array_t<bool> &arr = pybind11::cast<pybind11::array_t<bool>>(mask);
+
+            if (arr.ndim() != 2) {
+                throw std::invalid_argument(
+                "broadcast_pauli_errors can only accept mask that is a 2D array of np.bool_");
+            }
+
+            size_t major = arr.shape(0);
+            size_t minor = arr.shape(1);
+            auto u = arr.unchecked<2>();
```

Comment:
You didn't check that the size of the array is correct.

For the qubit axis, you should be lenient. This is consistent with the simulator auto-expanding when new qubits are mentioned. Call `self.ensure_safe_to_do_circuit_with_stats(CircuitStats{.num_qubits=X})` to ensure the simulation state is not too small. Continue to iterate up to the array's qubit count, which may be less than the simulator's qubit count.

For the instance count, you should be strict. This is where the speed comes from. If the sim's batch size differs from the array shape, throw an exception.